### PR TITLE
Maria - Tigers

### PR DESCRIPTION
**Pull request by mariasilvadev**

### CODE REVIEW

**Comment by mariasilvadev** [comment]
Completed project

File: tests/test_wave_01.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 from viewing_party.party import *
 from tests.test_constants import *
 
-@pytest.mark.skip()
+# @pytest.mark.skip()
```
Comment (chimerror) [inline tests/test_wave_01.py:7]: Minor style preference: I would just delete these skip decorators rather than comment them out. Just helps keep the code cleaner.

There's a few other skip decorators that are like this, but I'll only say it once.

Having talked to other students, they ran into issues trying to delete rather than comment, but I tried it myself and it seemed to work OK.

File: tests/test_wave_01.py [hunk 2]
```diff
@@ -119,15 +119,16 @@ def test_moves_movie_from_watchlist_to_empty_watched():
     assert len(updated_data["watchlist"]) == 0
     assert len(updated_data["watched"]) == 1
     
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
```
Comment (chimerror) [inline tests/test_wave_01.py:122]: As a minor style preference, I recommend just getting rid of commented out code in general unless there's a clear reason to keep it, which should be explained in a comment. This also applies to the comments below where we described how we wanted you to extend the tests, but more just in the sense of keeping code clean.

Deleting these comments would have the additional benefit of keeping the `assert` on line 126 closer to the other asserts in this function.

I'm sure there's a few other cases where this applies, but I'll only point it out once to avoid repeating myself.

File: tests/test_wave_01.py [hunk 3]
```diff
@@ -119,15 +119,16 @@ def test_moves_movie_from_watchlist_to_empty_watched():
...
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
-
-@pytest.mark.skip()
+    assert updated_data["watched"][0]["title"] == MOVIE_TITLE_1
```
Comment (chimerror) [inline tests/test_wave_01.py:126]: This test is OK as a quick check, but as a minor test design thing, I would recommend checking all the keys of the movie that got watched, not just the title. The easiest way to do this would likely to save the expected dictionary and use `==` to compare the entire dictionary, but admittedly, the arrange part of this code that we wrote would have to change...

File: tests/test_wave_01.py [hunk 4]
```diff
@@ -143,12 +144,12 @@ def test_moves_movie_from_watchlist_to_watched():
...
     assert len(updated_data["watched"]) == 2
     
-    raise Exception("Test needs to be completed.")
+    # raise Exception("Test needs to be completed.")
     # *******************************************************************************************
     # ****** Add assertions here to test that the correct movie was added to "watched" **********
     # *******************************************************************************************
-
-@pytest.mark.skip()
+    assert updated_data["watched"][1] == HORROR_1
```
Comment (chimerror) [inline tests/test_wave_01.py:151]: Following on the last test design issue, yes, this is a good way to check all keys, not just the title, so good work!

File: viewing_party/party.py [hunk 5]
```diff
@@ -1,23 +1,162 @@
 # ------------- WAVE 1 --------------------
+from platform import libc_ver
```
Comment (chimerror) [inline viewing_party/party.py:2]: Looks like this `import` statement never got used. I know there's a VS Code extension that will add these automatically that often causes it by offering suggestions that create these imports but are most likely not what you want. I would recommend disabling it, Ansel describes how to in [this Slack post](https://adad18.slack.com/archives/C03QJHJDW2V/p1663446293671619?thread_ts=1663446288.654079&cid=C03QJHJDW2V).

Failing that, make sure to remove these types of statements if they never get used just as a style practice.

File: viewing_party/party.py [hunk 6]
```diff
@@ -1,23 +1,162 @@
...
+            new_movie["genre"] = genre
+            if rating != None:
+                new_movie["rating"] = rating
+                return new_movie
+            else:
+                return None
+        else:
+            return None
+    else:
+        return None
```
Comment (chimerror) [inline viewing_party/party.py:19]: This code works, but in general if you find yourself nesting code so deeply, it's likely a sign that there is a better way to arrange the code. About more than 3 levels of nesting becomes hard to read.

Towards that end, I would consider moving them to check the input arguments before you go ahead and initialize `new_movie`. That way, if any are `None`, you can avoid the allocation of `new_movie`.

Additionally, you can definitely get this down to one if statement either by using `and` or putting them into a temporary list like this:

```python
if None in [title, genre, rating]:
    return None
```

Now, while this is more succinct, it could easily be argued that separate checks are more explicit. They also would make it easier to expand for other checks of the individual values:

```python
if rating == None or rating < 0 or rating > 5:
    return None
```

But this is a rather minor issue.

File: viewing_party/party.py [hunk 7]
```diff
@@ -1,23 +1,162 @@
...
+    user_data["watchlist"].append(movie)
+    return user_data
+
+def watch_movie(user_data, title):
+    for i in range(len(user_data["watchlist"])):
+        if title == user_data["watchlist"][i]["title"]: 
+            user_data["watched"].append(user_data["watchlist"][i])
+            user_data["watchlist"].pop(i)
+            
+            return user_data
```
Comment (chimerror) [inline viewing_party/party.py:36]: While using `range` and `pop` here is fine, you can definitely make this a bit cleaner and easier to read by using the `for` loop to iterate through the movies in `user_data["watchlist"]` directly, though you'll need to replace the `pop` call with `remove`:

```python
for movie in user_data["watchlist"]:
    if title == movie["title"]:
        user_data["watched"].append(movie)
        user_data["watchlist"].remove(movie)
        return user_data
```

Comment (chimerror) [reply viewing_party/party.py:36]: However, good job at avoiding the issue of modifying the list you are iterating through by returning early on line 36!

File: viewing_party/party.py [hunk 8]
```diff
@@ -1,23 +1,162 @@
...
+    return user_data
+
+
```
Comment (chimerror) [inline viewing_party/party.py:39]: Very minor style nitpick: Keep an eye out for unneeded blank lines like these. Yeah, they don't really hurt anything but it's part of maintaining a consistent style. I would just delete the extra one(s) until there's only one blank line.

In some style preferences, two blank lines are used to provide better separation between functions or classes, but as you don't seem to be doing that elsewhere, I'm going to assume you're not sticking with that, and thus point it out.

File: viewing_party/party.py (hunk 7 above)
Comment (chimerror) [reply viewing_party/party.py:36]: There are other loops that could be cleaned up by using `for` like this rather than `range` but I won't point them all out in the goal of avoiding repeating myself.

File: viewing_party/party.py [hunk 9]
```diff
@@ -1,23 +1,162 @@
...
+        
+        counter = 0
+        most_frequent_genre = genre_list[0]
+        for i in genre_list:
+            curr_frequency = genre_list.count(i)
+            if(curr_frequency> counter):
+                counter = curr_frequency
+                most_frequent_genre = i
+ 
+        return most_frequent_genre
```
Comment (chimerror) [inline viewing_party/party.py:67]: This implementation works, but does a bit more work than necessary. Since `for i in genre_list` will go through all elements in `genre_list`, that means it will check multiple times if there are repeated genres.

While y'all hadn't been introduced to Big O notation when you wrote this, there is a performance downside that can be highlighted as such. The `for` loop is going to go through the n genres in `genre_list`, where n is the number of movies in `user_data["watched"]`, which is O(n). For each of these n genres, it will call `genre_list.count(i)`, which is itself O(n). Combined this makes it O(n^2).

Rather than making a list of the genres, you can use a dictionary to keep track of the frequencies of the genres and find the answer in O(n) time. You're not required to do so, but I do recommend trying to implement that answer as it is a very common pattern.

File: viewing_party/party.py [hunk 10]
```diff
@@ -1,23 +1,162 @@
...
+        return 0.0
 
-
+def get_most_watched_genre(user_data):
+    if len(user_data["watched"]) >= 1:
+        genre_list = []
+        for i in range(len(user_data["watched"])):
+            genre_list.append(user_data["watched"][i]["genre"])
+        
+        counter = 0
```
Comment (chimerror) [inline viewing_party/party.py:59]: Minor naming issue: `counter` as a name implies that something is going to be continually incremented to keep count, but in this case, `counter` is more keeping track of the current maximum count of movies in a genre. As such, a name like `max_genre_count` would be more clear.

File: viewing_party/party.py [hunk 11]
```diff
@@ -1,23 +1,162 @@
...
+        most_frequent_genre = genre_list[0]
+        for i in genre_list:
```
Comment (chimerror) [inline viewing_party/party.py:61]: Minor naming issue: Single letter variables like `i` are usually reserved for keeping track of indices as a list is iterated through. However, in this case, we are not iterating through the indices, but the actual genres themselves. As such, a name like `genre` would be more clear.

File: viewing_party/party.py [hunk 12]
```diff
@@ -1,23 +1,162 @@
...
+        return most_frequent_genre
+    else:
+        return None
 # -----------------------------------------
 # ------------- WAVE 3 --------------------
 # -----------------------------------------
+def get_unique_watched(user_data):
+    my_movies_watched = []
+    for movie in user_data["watched"]:
+        my_movies_watched.append(movie)
```
Comment (chimerror) [inline viewing_party/party.py:76]: I'm not sure that copying all of `user_data["watched"]` into `my_movies_watched` here provides much benefit over just iterating over `user_data["watched"]`. By copying, new memory is allocated for the new list `my_movies_watched`.

Even though using local variables to save particularly complex statements is a good way to make code easier to read, that can be done without the copy:

```python
my_movies_watched = user_data["watched"]
```

File: viewing_party/party.py [hunk 13]
```diff
@@ -1,23 +1,162 @@
...
+    
+    friend_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.append(movie)
+
+    unique_movies = []
 
+    for movie in my_movies_watched:
+        if movie not in friend_movies and movie not in unique_movies:
```
Comment (chimerror) [inline viewing_party/party.py:86]: Interestingly, you don't need the `movie not in unique_movies` check here if we can assume that there are no repeated movies in `my_movies_watched`. As it is a copy of `user_data["watched"]` that seems like a fair assumption to make.

However, of course multiple friends may watch the same movie so a check like this would be useful in the `for` loop above to de-duplicate the friends' movies. (It's only a minor performance hit that you didn't do this, so no worries.)

File: viewing_party/party.py (hunk 12 above)
Comment (chimerror) [reply viewing_party/party.py:76]: It looks like you did the same in `get_friends_unique_watched`, and possibly other places, but I'll only point it out once.

File: viewing_party/party.py [hunk 14]
```diff
@@ -1,23 +1,162 @@
...
+    
+    friend_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.append(movie)
+
+    unique_movies = []
+
+    for movie in friend_movies:
+        if movie not in my_movies_watched and movie not in unique_movies:
```
Comment (chimerror) [inline viewing_party/party.py:104]: Unlike `get_unique_watched`, the `movie not in unique_movies` check _is_ necessary here, as multiple friends can watch the same movie.

I'm sure that seems obvious, but I just wanted to make sure you didn't take my advice from `get_unique_watched` as applying here too.

File: viewing_party/party.py [hunk 15]
```diff
@@ -1,23 +1,162 @@
...
+def get_available_recs(user_data):
+    recommended_movies = []
+    my_movies_watched = []
+    for movie in user_data["watched"]:
+        my_movies_watched.append(movie)
+    
+    friend_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.append(movie)
```
Comment (chimerror) [inline viewing_party/party.py:120]: Make sure to take opportunities to reuse earlier waves' code as helper functions. For example, all this code duplicates exactly what was done in `get_friends_unique_watched`, so it could be used as a helper function:

```python
friends_unique_watched = get_friends_unique_watched(user_data)

recommended_movies = []
for movie in friends_unique_watched:
    if movie["host"] in user_data["subscriptions"]:
        recommended_movies.append(movie)

return recommended_movies
```

File: viewing_party/party.py [hunk 16]
```diff
@@ -1,23 +1,162 @@
...
+    for movie in friend_movies:
+        if movie not in my_movies_watched and movie not in unique_movies:
+            unique_movies.append(movie)
+        
+    return unique_movies
 # -----------------------------------------
 # ------------- WAVE 4 --------------------
 # -----------------------------------------
+def get_available_recs(user_data):
+    recommended_movies = []
```
Comment (chimerror) [inline viewing_party/party.py:112]: So minor style nitpick, Python has no requirement that all local variables be defined at the top of the function, so I usually wait until the last possible moment to define them. There are two benefits to this. 

First is a chance to avoid unnecessary memory allocation. It doesn't really apply in this case, but if you had checks that made an early return before you actually ended up needing `recommended_movies`, you can avoid the memory allocation for `recommended_movies`. Not a big deal for an empty list, but it could be with larger data structures or objects that open other resources such as network connections.

The second is more on the readability side. By defining `recommended_movies` right when it gets filled in, it more obviously ties `recommended_movies` with the `for` loop that fills it. In your current code, it's not too far away, but if you imagine you have more checks of the passed-in values or other code for other tasks, it becomes easier to forget what `recommended_movies` started as.

File: viewing_party/party.py [hunk 17]
```diff
@@ -1,23 +1,162 @@
...
 # -----------------------------------------
+def get_new_rec_by_genre(user_data):
+    my_movies_watched = []
+    for movie in user_data["watched"]:
+        my_movies_watched.append(movie)
+    
+    friend_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.append(movie)
```
Comment (chimerror) [inline viewing_party/party.py:140]: Just like `get_available_recs` in wave 4, you could have reused `get_friends_unique_watched` as a helper function and made this code a bit cleaner.

File: viewing_party/party.py [hunk 18]
```diff
@@ -1,23 +1,162 @@
...
+
+    recommended_by_genre = []
+
+    user_popular_genre = get_most_watched_genre(user_data)
```
Comment (chimerror) [inline viewing_party/party.py:144]: Though good job at reusing `get_most_watched_genre` as a helper function here!

File: viewing_party/party.py [hunk 19]
```diff
@@ -1,23 +1,162 @@
...
+    friend_movies = []
+    for friend in user_data["friends"]:
+        for movie in friend["watched"]:
+            friend_movies.append(movie)
+    from_favorite_recommendation = []
+    for movie in user_data["favorites"]:
+        if movie not in friend_movies and movie not in from_favorite_recommendation:
+            from_favorite_recommendation.append(movie)
+            
+    return from_favorite_recommendation
```
Comment (chimerror) [inline viewing_party/party.py:162]: In this case, however, an argument can be made that there might be movies in `user_data["favorites"]` that are not in `user_data["watched"]` and thus `get_friends_unique_watched` can't be reused.

I think it's reasonable to make the assumption that a user would not favorite a movie without watching it, so I think it would have been fine to reuse `get_friends_unique_watched` here too.

Comment (chimerror) [reply viewing_party/party.py:162]: Though, as a minor style issue, please do use blank lines to separate out "sections" of your code. It can be very hard to follow code without them, and even lead to mistakes in indentation.

You generally did this elsewhere, so I am not sure why not here.

**Review by chimerror — COMMENTED** [review]
Good Work!

While I point out several minor issues on a variety of topics, this code is functional and generally readable, and thus demonstrates sufficient understanding of the concepts covered. As such, I am marking this Green.
